Avoid creating TrainerConfig during parse_conf_file()

emailweixu opened the pull request:
Before this change, get_observation() will create a TrainerConfig(). This makes the configurable
values of TrainerConfig to be all labelled as used. And it will not be allowed to change after get_observation().

Now we use get_config() to get the value of TrainerConfig.data_transformer_ctor
to avoid creating TrainerConfig object. So the configs of other TrainerConfig values
can still be changed after get_observation_spec().

Also changed sac_carla.gin to use alf-config

emailweixu committed:
Avoid creating TrainerConfig during parse_conf_file()

Before this change, get_observation() will create a TrainerConfig(). This makes the configurable
values of TrainerConfig to be all labelled as used. And it will not be allowed to change after get_observation().

Now we use get_config() to get the value of TrainerConfig.data_transformer_ctor
to avoid creating TrainerConfig object. So the configs of other TrainerConfig values
can still be changed after get_observation_spec().

Also changed sac_carla.gin to use alf-config

hnyu commented on alf/environments/suite_carla.py:
```diff
@@ -1027,7 +1027,7 @@ def __init__(self,
             batch_size (int): the number of learning vehicles.
             map_name (str): the name of the map (e.g. "Town01")
-            vehicle_filter (str): the filter for getting vehicle blueprints.
+            vehicle_filter (str): the filter for getting ego vehicle blueprints.
```
I think 'ego' is a little confusing here. Might just explain that this filter is used for vehicles that will be trained to drive. And the other vehicles will use a catch-all filter.

emailweixu replied:
Updated.

le-horizon commented on alf/bin/train.py:
```diff
@@ -87,9 +85,17 @@ def train_eval(ml_type, root_dir):
-    common.parse_conf_file(conf_file)
-    train_eval(FLAGS.ml_type, FLAGS.root_dir)
+    try:
+        common.parse_conf_file(conf_file)
```
will it raise error if parse_conf_file tries to set random seed again?

emailweixu replied:
moved all random_seed ralated code to get_env

le-horizon replied:
Looks nice!

le-horizon commented on alf/bin/train.py:
```diff
@@ -87,9 +85,17 @@ def train_eval(ml_type, root_dir):
 
 def main(_):
     FLAGS.alsologtostderr = True
+    random_seed = alf.get_config('TrainerConfig.random_seed')
```
will this only read from the command line config?  What's the reason for making this change: first get random_seed, then set it and set config, then parse conf file?  It seems complicated and prone to error?

emailweixu replied:
moved to config_helpers.get_env() and commented there.

le-horizon reviewed (commented):
Some values need to be configured before parse_alf_config, so the config and initialization of alf seems intertwined, and a bit hard to follow.  We may need to create some documentation for this, in case we forget in the future?

Haichao-Zhang commented on alf/config_util.py:
```diff
@@ -189,20 +194,8 @@ def is_used(self):
-
-    """
+def _get_config_node(config_name):
+    """Get the _Config corresponding to config_name."""
```
Does ```_Config``` has any special meaning here? Maybe can change to ```config node```.

emailweixu replied:
_Config is the class name

Haichao-Zhang commented on alf/config_util.py:
```diff
@@ -224,7 +217,26 @@ def config1(config_name, value, mutable=True):
+    """
+    config_node = _get_config_node(config_name)
+
+    if raise_if_used and config_node.is_used():
```
When raise_if_used is False,  maybe we can generate a warning message

emailweixu replied:
With the current way using raise_if_used=True inside config_helper, the config is always used, and a warning message will be quite confusing.

Haichao-Zhang commented on alf/config_util.py:
```diff
@@ -250,6 +262,28 @@ def config1(config_name, value, mutable=True):
         config_node.set_mutable(mutable)
 
 
+def get_config(config_name):
```
Maybe we can change the name of this function to ```get_config_value```?

emailweixu replied:
changed

emailweixu committed:
parse_conf_file always create global env and init seed

le-horizon commented on alf/examples/sac_carla_conf.py:
```diff
@@ -0,0 +1,206 @@
+    sparse_reward=False,
+    # Cannot train well with allow_negative_distance_reward=False yet.
+    allow_negative_distance_reward=True,
+    # Currently, even a small penalty such as one make the training much worse
```
make->makes

emailweixu replied:
removed this comment as it is not accurate anymore

le-horizon commented on alf/examples/ac_cart_pole_conf.py:
```diff
@@ -0,0 +1,53 @@
+alf.config('ValueNetwork', fc_layer_params=(100, ))
+alf.config(
+    'ActorCriticAlgorithm',
+    optimizer=alf.optimizers.Adam(lr=1e-3, gradient_clipping=10.0))
```
It's fine as is.  Does it make sense to have lr configured using a same named variable in TrainerConfig?

emailweixu replied:
There could be multiple optimizers.

le-horizon commented on alf/config_helpers.py:
```diff
@@ -94,19 +99,71 @@ def get_action_spec():
+        else:
+            # We construct a TrainerConfig object here so that the value
+            # configured through gin-config can be properly retrieved.
+            train_config = TrainerConfig(root_dir='')
```
I'm confused.  This config_helper.py is only used by alf config, not gin-config?  Why do we need to care about gin-config here?

emailweixu replied:
Because get_env() will be used by policy_trainer to get the global env even when gin-config is used for configuration.

le-horizon commented on alf/examples/sac_carla_conf.py:
```diff
@@ -0,0 +1,206 @@
+    joint_fc_layer_params=fc_layers_params,
+    activation=activation,
+    use_fc_bn=use_batch_normalization,
+    output_tensor_spec=env.reward_spec())
```
maybe have a common.get_reward_spec() function, and hide alf.get_env() inside it?

emailweixu replied:
There are many other env member function. It is not a good idea to do this for all of them. The reason of having get_observation_spec() is that it has additional handling of data_transformer.

le-horizon reviewed (commented):
Accidentally submitted the partial review.
Also took a look at the new commit.

It's much cleaner now, only a few nits/suggestions if they make sense.

Thanks,
Le

emailweixu committed:
Better handling of --conf_param

Previously, if the related module has not been imported, the config
name specified through --conf_param will not be found. This make it
very hard to use since the related modules are typically imported
in the config file. This change delay the binding of conf_param
until the related module has been loaded.

emailweixu committed:
Update comments of sac_carla_conf.py

le-horizon reviewed (approved):
LG.
